latch_cli/services/local_dev.py

```diff
@@ -0,0 +1,319 @@
+    # see https://github.com/aaugustin/websockets/issues/867
+    await asyncio.sleep(0)
+
+
+async def run_local_dev_session(pkg_root: Path):
+    scripts_dir = pkg_root.joinpath("scripts").resolve()
+
+    def file_filter(filename: str) -> bool:
+        file_path = Path(filename).resolve()
+        return file_path == scripts_dir or scripts_dir in file_path.parents
+
+    completer = NestedCompleter(
+        {
+            "run-script": PathCompleter(
+                get_paths=lambda: [str(pkg_root)],
+                file_filter=file_filter,
+            ),
+            "run": None,
+            "ls": None,
+            "list-tasks": None,
+            "shell": None,
+        },
+        ignore_case=True,
+    )
+    session = PromptSession(
+        ">>> ",
+        completer=completer,
+        complete_while_typing=True,
+        auto_suggest=AutoSuggestFromHistory(),
+        history=FileHistory(pkg_root.joinpath(".latch_history")),
+    )
+
+    key_path = pkg_root / ".ssh_key"
+
+    with TemporarySSHCredentials(key_path) as ssh:
+        headers = {"Authorization": f"Bearer {retrieve_or_login()}"}
+
+        cent_resp = post(
+            sdk_endpoints["provision-centromere"],
+            headers=headers,
+            json={"public_key": ssh.public_key},
+        )
+
+        init_local_dev_resp = post(
+            sdk_endpoints["local-development"],
+            headers=headers,
+            json={
+                "ws_account_id": current_workspace(),
+                "pkg_root": pkg_root.name,
+            },
+        )
+
+        if init_local_dev_resp.status_code == 403:
+            raise ValueError("You are not authorized to use this feature.")
+
+        cent_data = cent_resp.json()
+        centromere_ip = cent_data["ip"]
+        centromere_username = cent_data["username"]
+
+        init_local_dev_data = init_local_dev_resp.json()
+        access_key = init_local_dev_data["docker"]["tmp_access_key"]
+        secret_key = init_local_dev_data["docker"]["tmp_secret_key"]
+        session_token = init_local_dev_data["docker"]["tmp_session_token"]
+        port = init_local_dev_data["port"]
+
+        # ecr setup
+        ws_id = current_workspace()
+        if int(ws_id) < 10:
+            ws_id = f"x{ws_id}"
+
+        try:
+            ecr_client = boto3.session.Session(
+                aws_access_key_id=access_key,
+                aws_secret_access_key=secret_key,
+                aws_session_token=session_token,
+                region_name="us-west-2",
+            ).client("ecr")
+
+            ecr_auth = ecr_client.get_authorization_token()
+            docker_access_token = ecr_auth["authorizationData"][0]["authorizationToken"]
+
+        except Exception as err:
+            raise ValueError(f"unable to retrieve an ecr login token") from err
+
+        try:
+            paginator = ecr_client.get_paginator("describe_images")
+            image_name = f"{ws_id}_{pkg_root.name}"
```

Comment:
the image name is now in two places (original register code as well), we should fix that

Reply:
adding a todo, but this will switch to an endpoint when i figure the different container per task dilemma